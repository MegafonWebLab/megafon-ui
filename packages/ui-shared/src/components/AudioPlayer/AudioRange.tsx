import React from 'react';
import { cnCreate, filterDataAttrs } from '@megafon/ui-helpers';
import './style/AudioRange.less';

export interface IAudioRangeProps {
    className?: string;
    maxValue: number | string;
    value: number;
    step: string;
    colorPercent: number;
    onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
    onMouseUp?: () => void;
    onTouchEnd?: () => void;
    dataAttrs?: Record<string, string>;
}

const cn = cnCreate('mfui-audio-range');
const AudioRange: React.FC<IAudioRangeProps> = ({
    className,
    maxValue,
    value,
    step,
    colorPercent,
    onChange,
    onMouseUp,
    onTouchEnd,
    dataAttrs,
}) => (
    <input
        className={cn([className])}
        style={{
            backgroundSize: `${colorPercent}% 100%`,
        }}
        type="range"
        min="0"
        max={maxValue}
        step={step}
        value={value}
        onChange={onChange}
        onMouseUp={onMouseUp}
        onTouchEnd={onTouchEnd}
        {...filterDataAttrs(dataAttrs)}
    />
);

export default AudioRange;
