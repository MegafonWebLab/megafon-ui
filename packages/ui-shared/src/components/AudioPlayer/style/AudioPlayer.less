@import '~@megafon/ui-core/styles/base.less';

@b: e('.mfui-audio-player');

@{b} {
    margin-top: 32px;

    &__player {
        display: flex;
        align-items: center;
        justify-content: space-between;
        box-sizing: border-box;
        padding: 16px;
        border-radius: 12px;

        background: var(--spbSky0);

        @media @mobileB {
            max-width: 530px;
        }
        @media @desktop {
            max-width: 550px;
        }
    }

    &__playing-controls {
        margin-right: 16px;
    }
}

// mods position

@{b}_position_center {
    @{b}__player {
        margin: 0 auto;
    }
}

@{b}_position_left {
    @{b}__player {
        margin-right: auto;
    }
}

@{b}_position_right {
    @{b}__player {
        margin-left: auto;
    }
}

// mods fullWidth

@{b}_full-width {
    @{b}__player {
        @media @mobileB {
            max-width: 100%;
        }
        @media @desktop {
            max-width: 100%;
        }
    }
}
