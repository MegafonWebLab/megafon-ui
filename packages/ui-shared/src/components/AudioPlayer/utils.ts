const getMinutes = (sec: number): string => {
    const min = parseInt(`${sec / 60}`, 10);
    if (min < 10) {
        return `0${min}`;
    }

    return `${min}`;
};

/**
 * @param {Number} sec
 * @returns {String}
 */
const getSeconds = (sec: number): string => {
    if (sec < 10) {
        return `0${sec}`;
    }

    return `${sec}`;
};

export const timerFormat = (sec: number): string => {
    const min = getMinutes(sec);
    const seconds = getSeconds(sec % 60);

    return `${min}:${seconds}`;
};
