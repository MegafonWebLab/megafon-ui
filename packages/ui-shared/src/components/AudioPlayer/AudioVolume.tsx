import React from 'react';
import { Tooltip } from '@megafon/ui-core';
import { cnCreate } from '@megafon/ui-helpers';
import AudioRange from './AudioRange';
import NoVolumeIcon from './i/no-volume.svg';
import VolumeIcon from './i/volume.svg';
import './style/AudioVolume.less';

export interface IAudioVolumeProps {
    onChangeAudioVolume?: (volume: number) => void;
}

const cn = cnCreate('mfui-audio-volume');
const AudioVolume: React.FC<IAudioVolumeProps> = ({ onChangeAudioVolume }) => {
    const tooltipTrigger = React.useRef<HTMLDivElement | null>(null);

    const [trackVolume, setTrackVolume] = React.useState<number>(0.3);

    const volumeColorPercent: number = React.useMemo(() => trackVolume * 100, [trackVolume]);

    const handleScrubVolume = (e: React.ChangeEvent<HTMLInputElement>) => {
        const currentVolume = Number(e.target.value);

        setTrackVolume(currentVolume);
        onChangeAudioVolume?.(currentVolume);
    };

    return (
        <div className={cn()}>
            <div className={cn('icon-area')} ref={tooltipTrigger}>
                <VolumeIcon className={cn('icon', { visible: !!trackVolume })} />
                <NoVolumeIcon className={cn('icon', { visible: !trackVolume })} />
            </div>
            <Tooltip
                classes={{
                    content: `${cn('tooltip-content')}`,
                    contentShadow: `${cn('tooltip-content-shadow')}`,
                }}
                triggerElement={tooltipTrigger}
                triggerEvent="click"
                placement="top"
                paddings="none"
            >
                <AudioRange
                    className={cn('range')}
                    maxValue="1"
                    value={trackVolume}
                    step="0.01"
                    colorPercent={volumeColorPercent || 0}
                    onChange={handleScrubVolume}
                    dataAttrs={{ 'data-testid': 'AudioVolumeRange' }}
                />
            </Tooltip>
        </div>
    );
};

export default AudioVolume;
