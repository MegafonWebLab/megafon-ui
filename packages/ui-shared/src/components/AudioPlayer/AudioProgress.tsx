import React from 'react';
import { cnCreate } from '@megafon/ui-helpers';
import AudioRange from './AudioRange';
import { timerFormat } from './utils';
import './style/AudioProgress.less';

export const INTERVAL_DELAY = 1000;

export interface IAudioProgressProps {
    audioRef: React.MutableRefObject<HTMLAudioElement | null>;
    audioTitle: string;
    isPlaying: boolean;
    isPause: boolean;
    onChangeAudioCurrentTime: (currentTime: number) => void;
    onPLay: () => void;
    onSetIsPlaying: (value: boolean) => void;
}

const cn = cnCreate('mfui-audio-progress');
const AudioProgress: React.FC<IAudioProgressProps> = ({
    audioRef,
    audioTitle,
    isPlaying,
    isPause,
    onChangeAudioCurrentTime,
    onPLay,
    onSetIsPlaying,
}) => {
    const intervalId = React.useRef<NodeJS.Timeout | null>(null);

    const [trackDuration, setTrackDuration] = React.useState<number>(0);
    const [trackProgress, setTrackProgress] = React.useState<number>(0);

    const progressColorPercent: number = React.useMemo(
        () => (trackProgress / trackDuration) * 100,
        [trackDuration, trackProgress],
    );
    const progressValue: number = React.useMemo(
        () => (isPlaying || isPause ? trackProgress : trackDuration),
        [isPause, isPlaying, trackDuration, trackProgress],
    );

    const handleStartTimer = React.useCallback(() => {
        intervalId.current && clearInterval(intervalId.current);

        intervalId.current = setInterval(() => {
            if (audioRef.current?.ended) {
                onSetIsPlaying(false);
                setTrackProgress(Math.floor(audioRef.current?.duration || 0));

                return;
            }

            setTrackProgress(Math.floor(audioRef.current?.currentTime || 0));
        }, INTERVAL_DELAY);
    }, [audioRef, onSetIsPlaying]);

    const handleScrubProgress = (e: React.ChangeEvent<HTMLInputElement>) => {
        const currentTime = Number(e.target.value);

        setTrackProgress(currentTime);
        onChangeAudioCurrentTime(currentTime);
    };

    const handleScrubEndProgress = () => {
        if (!isPlaying) {
            onPLay();
        }

        handleStartTimer();
    };

    React.useEffect(() => {
        setTrackProgress(Math.floor(audioRef?.current?.currentTime || 0));
    }, [audioRef]);

    React.useEffect(() => {
        const audioNode = audioRef.current;

        if (audioNode) {
            audioNode.onloadedmetadata = () => {
                setTrackDuration(Math.round(audioNode.duration));
            };
        }
    }, [audioRef]);

    React.useEffect(() => {
        if (isPlaying) {
            handleStartTimer();
        }

        if (isPause) {
            intervalId.current && clearInterval(intervalId.current);
        }
    }, [handleStartTimer, isPause, isPlaying]);

    React.useEffect(
        () => (): void => {
            intervalId.current && clearInterval(intervalId.current);
        },
        [],
    );

    return (
        <div className={cn()}>
            <div className={cn('info')}>
                <span className={cn('title')}>{audioTitle}</span>
                <span className={cn('value')}>{timerFormat(progressValue)}</span>
            </div>
            <div className={cn('range-wrapper')}>
                <AudioRange
                    className={cn('range')}
                    maxValue={trackDuration}
                    value={trackProgress}
                    step="1"
                    colorPercent={progressColorPercent || 0}
                    onChange={handleScrubProgress}
                    onMouseUp={handleScrubEndProgress}
                    onTouchEnd={handleScrubEndProgress}
                />
            </div>
        </div>
    );
};

export default AudioProgress;
