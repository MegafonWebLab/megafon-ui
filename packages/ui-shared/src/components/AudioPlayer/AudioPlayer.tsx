import React from 'react';
import { Button } from '@megafon/ui-core';
import { cnCreate } from '@megafon/ui-helpers';
import AudioProgress from './AudioProgress';
import AudioVolume from './AudioVolume';
import IconPause from './i/pause.svg';
import IconPlay from './i/play.svg';
import './style/AudioPlayer.less';

export interface IAudioPlayerProps {
    audioSrc: string;
    audioTitle: string;
    position?: 'left' | 'center' | 'right';
    isFullWidth?: boolean;
}

const cn = cnCreate('mfui-audio-player');
const AudioPlayer: React.FC<IAudioPlayerProps> = ({ audioSrc, audioTitle, position = 'center', isFullWidth }) => {
    const audioRef = React.useRef<HTMLAudioElement | null>(null);

    const [isPlaying, setIsPlaying] = React.useState<boolean>(false);
    const [isPause, setIsPause] = React.useState<boolean>(false);

    const handlePLay = () => {
        setIsPlaying(true);
        setIsPause(false);

        audioRef.current?.play();
    };

    const handlePause = () => {
        setIsPlaying(false);
        setIsPause(true);

        audioRef.current?.pause();
    };

    const handleChangeAudioCurrentTime = (currentTime: number) => {
        if (!audioRef || !audioRef.current) {
            return;
        }

        audioRef.current.currentTime = currentTime;
    };

    const handleChangeAudioVolume = (volume: number) => {
        if (!audioRef || !audioRef.current) {
            return;
        }

        audioRef.current.volume = volume;
        audioRef.current.muted = !volume;
    };

    const playingIconHandleClick: () => void = React.useMemo(() => (isPlaying ? handlePause : handlePLay), [isPlaying]);

    return (
        <div className={cn({ 'full-width': isFullWidth, position })}>
            <div className={cn('player')}>
                {/* eslint-disable-next-line jsx-a11y/media-has-caption */}
                <audio className={cn('audio')} ref={audioRef} controls={false}>
                    <source src={audioSrc} type="audio/mpeg" />
                </audio>
                <div className={cn('playing-controls')}>
                    <Button
                        icon={isPlaying ? <IconPause /> : <IconPlay />}
                        onClick={playingIconHandleClick}
                        sizeAll="small"
                        className={cn('button')}
                    />
                </div>
                <AudioProgress
                    audioRef={audioRef}
                    audioTitle={audioTitle}
                    isPlaying={isPlaying}
                    isPause={isPause}
                    onChangeAudioCurrentTime={handleChangeAudioCurrentTime}
                    onPLay={handlePLay}
                    onSetIsPlaying={setIsPlaying}
                />
                <AudioVolume onChangeAudioVolume={handleChangeAudioVolume} />
            </div>
        </div>
    );
};

export default AudioPlayer;
