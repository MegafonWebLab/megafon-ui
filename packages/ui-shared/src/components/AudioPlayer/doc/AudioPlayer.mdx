---
name: AudioPlayer
menu: Конструктор
route: components/audio-player
---

import { Props } from 'docz';
import AudioPlayer from '../AudioPlayer';
import Example from './AudioPlayer.example.mdx';

# AudioPlayer

<CopyToClipboardBox text="import { AudioPlayer } from '@megafon/ui-shared';" />

<DoczTabs>
    <Example />
    <Props of={AudioPlayer} />
</DoczTabs>
