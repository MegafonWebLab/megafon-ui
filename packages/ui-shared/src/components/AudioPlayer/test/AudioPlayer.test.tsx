import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import AudioPlayer, { IAudioPlayerProps } from '../AudioPlayer';

const requiredProps: IAudioPlayerProps = {
    audioSrc: 'test audioSrc',
    audioTitle: 'test audioTitle',
};

const playMock = jest.fn();
const pauseMock = jest.fn();

window.HTMLMediaElement.prototype.play = playMock;
window.HTMLMediaElement.prototype.pause = pauseMock;

jest.mock('@megafon/ui-core/dist/lib/components/Tooltip/Tooltip', () => ({ children }) => (
    <div data-name="Tooltip">{children}</div>
));

describe('AudioPlayer', () => {
    afterEach(() => jest.clearAllMocks());
    afterAll(() => jest.restoreAllMocks());

    it('should render component', () => {
        const { container } = render(<AudioPlayer {...requiredProps} />);

        expect(container).toMatchSnapshot();
    });

    it('should render with left position', () => {
        render(<AudioPlayer {...requiredProps} position="left" />);

        const rootEl = screen.getByTestId('AudioPlayer');

        expect(rootEl).toHaveClass('mfui-audio-player_position_left');
    });

    it('should render with full width property', () => {
        render(<AudioPlayer {...requiredProps} isFullWidth />);

        const rootEl = screen.getByTestId('AudioPlayer');

        expect(rootEl).toHaveClass('mfui-audio-player_full-width');
    });

    it('should play and pause audio after click on control button', () => {
        render(<AudioPlayer {...requiredProps} />);

        const btn = screen.getByTestId('AudioPlayer-btn');

        fireEvent.click(btn);
        expect(playMock).toBeCalled();

        fireEvent.click(btn);
        expect(pauseMock).toBeCalled();
    });

    it('should change audio current time', () => {
        render(<AudioPlayer {...requiredProps} />);

        const audioProgress = screen.getByTestId('AudioTimeRange');
        fireEvent.change(audioProgress, { target: { value: 50 } });

        const audio: HTMLMediaElement = screen.getByTestId('AudioPlayer-audio');
        expect(audio.currentTime).toBe(50);
    });

    it('should change audio volume', () => {
        render(<AudioPlayer {...requiredProps} />);

        const audioVolumeRange = screen.getByTestId('AudioVolumeRange');
        fireEvent.change(audioVolumeRange, { target: { value: 0.5 } });

        const audio: HTMLMediaElement = screen.getByTestId('AudioPlayer-audio');
        expect(audio.volume).toBe(0.5);
    });
});
