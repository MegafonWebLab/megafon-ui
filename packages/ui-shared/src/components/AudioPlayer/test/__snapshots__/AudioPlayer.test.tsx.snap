// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AudioPlayer /> snapshots should render component 1`] = `
<div
  className="mfui-audio-player mfui-audio-player_position_center"
>
  <div
    className="mfui-audio-player__player"
  >
    <audio
      className="mfui-audio-player__audio"
      controls={false}
    >
      <source
        src="test audioSrc"
        type="audio/mpeg"
      />
    </audio>
    <div
      className="mfui-audio-player__playing-controls"
    >
      <Button
        className="mfui-audio-player__button"
        icon={<test-file-stub />}
        onClick={[Function]}
        sizeAll="small"
      />
    </div>
    <AudioProgress
      audioRef={
        Object {
          "current": null,
        }
      }
      audioTitle="test audioTitle"
      isPause={false}
      isPlaying={false}
      onChangeAudioCurrentTime={[Function]}
      onPLay={[Function]}
      onSetIsPlaying={[MockFunction]}
    />
    <AudioVolume
      onChangeAudioVolume={[Function]}
    />
  </div>
</div>
`;

exports[`<AudioPlayer /> snapshots should render component when isPlaying 1`] = `
<div
  className="mfui-audio-player mfui-audio-player_position_center"
>
  <div
    className="mfui-audio-player__player"
  >
    <audio
      className="mfui-audio-player__audio"
      controls={false}
    >
      <source
        src="test audioSrc"
        type="audio/mpeg"
      />
    </audio>
    <div
      className="mfui-audio-player__playing-controls"
    >
      <Button
        className="mfui-audio-player__button"
        icon={<test-file-stub />}
        onClick={[Function]}
        sizeAll="small"
      />
    </div>
    <AudioProgress
      audioRef={
        Object {
          "current": null,
        }
      }
      audioTitle="test audioTitle"
      isPause={false}
      isPlaying={true}
      onChangeAudioCurrentTime={[Function]}
      onPLay={[Function]}
      onSetIsPlaying={[MockFunction]}
    />
    <AudioVolume
      onChangeAudioVolume={[Function]}
    />
  </div>
</div>
`;
